libraries/core_libs/consensus/src/pbft/pbft_manager.cpp

```diff
@@ -466,24 +466,27 @@ void PbftManager::resetPbftConsensus(uint64_t round) {
     executed_pbft_block_ = false;
   }
 
-  last_step_clock_initial_datetime_ = current_step_clock_initial_datetime_;
-  current_step_clock_initial_datetime_ = std::chrono::system_clock::now();
+  current_round_start_datetime_ = std::chrono::system_clock::now();
+}
+
+std::chrono::milliseconds PbftManager::elapsedTimeInMs(const time_point &start_time) {
+  return std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now() - start_time);
 }
 
 void PbftManager::sleep_() {
-  now_ = std::chrono::system_clock::now();
-  duration_ = now_ - round_clock_initial_datetime_;
-  elapsed_time_in_round_ms_ = std::chrono::duration_cast<std::chrono::milliseconds>(duration_).count();
-  LOG(log_tr_) << "elapsed time in round(ms): " << elapsed_time_in_round_ms_ << ", step " << step_;
-  // Add 25ms for practical reality that a thread will not stall for less than 10-25 ms...
-  if (next_step_time_ms_ > elapsed_time_in_round_ms_ + 25) {
-    auto time_to_sleep_for_ms = next_step_time_ms_ - elapsed_time_in_round_ms_;
-    LOG(log_tr_) << "Time to sleep(ms): " << time_to_sleep_for_ms << " in round " << getPbftRound() << ", step "
-                 << step_;
+  // Run "wait_for" sleep in loop due to potential spurious wakeup on lock
+  while (true) {
+    const auto round_elapsed_time = elapsedTimeInMs(current_round_start_datetime_);
+    if (next_step_time_ms_ <= round_elapsed_time) {
+      return;
+    }
+
+    const auto time_to_sleep_for_ms = next_step_time_ms_ - round_elapsed_time;
+    const auto [round, period] = getPbftRoundAndPeriod();
+    LOG(log_tr_) << "Sleep " << time_to_sleep_for_ms.count() << " [ms] before going into the next step. Period "
+                 << period << ", round " << round << ", step " << step_;
     std::unique_lock<std::mutex> lock(stop_mtx_);
-    stop_cv_.wait_for(lock, std::chrono::milliseconds(time_to_sleep_for_ms));
-  } else {
-    LOG(log_tr_) << "Skipping sleep, running late...";
+    stop_cv_.wait_for(lock, time_to_sleep_for_ms);
```

Comment:
I think you could use the return value of stop_cv_.wait_for to check if the time expired and return based on that

Reply:
return what ? This function is supposed to wait for time_to_sleep_for_ms. If wait_for exits before that due to spurious wake up, then what do you want to do ?